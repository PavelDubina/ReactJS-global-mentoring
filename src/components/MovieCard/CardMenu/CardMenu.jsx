import React, { useRef } from 'react'
import PropTypes from 'prop-types'
import { useOutsideClick } from '../../../Hooks/useClickOutside'
import styles from '../../../../styles/CardMenu.module.scss'

export const CardMenu = ({ isOpen, toggleMenu, onOpenModal }) => {
  const ref = useRef()
  useOutsideClick(ref, isOpen && toggleMenu)
  return (
    <>
      <div data-testid="card-menu" onClick={toggleMenu} className={styles.menu}>
        <div />
        <div />
        <div />
      </div>
      {isOpen && (
        <div ref={ref} className={styles.dropdown}>
          <div className={styles.dropdown_close} onClick={toggleMenu}>
            &#215;
          </div>
          <div className={styles.dropdown_item} onClick={onOpenModal}>
            Edit
          </div>
          <div className={styles.dropdown_item} onClick={onOpenModal}>
            Delete
          </div>
        </div>
      )}
    </>
  )
}

CardMenu.propTypes = {
  isOpen: PropTypes.bool.isRequired,
  toggleMenu: PropTypes.func.isRequired,
  onOpenModal: PropTypes.func.isRequired,
}
