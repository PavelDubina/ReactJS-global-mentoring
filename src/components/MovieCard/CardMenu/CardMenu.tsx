import React, { useRef } from 'react'
import { useTranslation } from 'next-i18next'
import { useOutsideClick } from '../../../Hooks/useClickOutside'
import styles from '../../../../styles/CardMenu.module.scss'

type CardMenuProps = {
  isOpen: boolean
  toggleMenu: () => void
  onOpenModal: (status: string) => void
}

export const CardMenu: React.FC<CardMenuProps> = ({ isOpen, toggleMenu, onOpenModal }) => {
  const { t } = useTranslation()
  const ref = useRef(null)
  const outsideFunc = isOpen ? toggleMenu : undefined
  useOutsideClick(ref, outsideFunc)
  return (
    <>
      <div data-testid="card-menu" onClick={toggleMenu} className={styles.menu}>
        <div />
        <div />
        <div />
      </div>
      {isOpen && (
        <div ref={ref} className={styles.dropdown}>
          <div className={styles.dropdown_close} onClick={toggleMenu}>
            &#215;
          </div>
          <div className={styles.dropdown_item} onClick={() => onOpenModal('edit')}>
            {t('menu.edit')}
          </div>
          <div className={styles.dropdown_item} onClick={() => onOpenModal('delete')}>
            {t('menu.delete')}
          </div>
        </div>
      )}
    </>
  )
}
