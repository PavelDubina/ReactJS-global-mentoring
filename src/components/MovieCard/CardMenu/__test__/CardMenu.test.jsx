import React from 'react'
import { fireEvent, render } from '@testing-library/react'
import { CardMenu } from '../CardMenu'

describe('CardMenu', () => {
  const getComponent = ({ toggleMenu = jest.fn(), isOpen = false, onOpenModal = jest.fn() }) =>
    render(<CardMenu toggleMenu={toggleMenu} isOpen={isOpen} onOpenModal={onOpenModal} />)
  it('Should renders correctly', () => {
    expect(getComponent({}).asFragment()).toMatchSnapshot()
  })
  it('Should renders correctly with opened dropdown', () => {
    expect(getComponent({ isOpen: true }).asFragment()).toMatchSnapshot()
  })
  it('Should called toggleMenu', () => {
    const toggleMenu = jest.fn()
    const { getByText } = getComponent({ toggleMenu, isOpen: true })
    fireEvent.click(getByText('Ã—'))
    expect(toggleMenu).toBeCalled()
  })
  it('Should called onOpenModal', () => {
    const onOpenModal = jest.fn()
    const { getByText } = getComponent({ onOpenModal, isOpen: true })
    fireEvent.click(getByText('Edit'))
    expect(onOpenModal).toBeCalled()
  })
})
