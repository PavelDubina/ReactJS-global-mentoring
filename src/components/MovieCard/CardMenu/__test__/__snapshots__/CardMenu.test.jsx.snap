// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CardMenu Should renders correctly 1`] = `
<DocumentFragment>
  <div
    class="menu"
    data-testid="card-menu"
  >
    <div />
    <div />
    <div />
  </div>
</DocumentFragment>
`;

exports[`CardMenu Should renders correctly with opened dropdown 1`] = `
<DocumentFragment>
  <div
    class="menu"
    data-testid="card-menu"
  >
    <div />
    <div />
    <div />
  </div>
  <div
    class="dropdown"
  >
    <div
      class="dropdown_close"
    >
      Ã—
    </div>
    <div
      class="dropdown_item"
    >
      Edit
    </div>
    <div
      class="dropdown_item"
    >
      Delete
    </div>
  </div>
</DocumentFragment>
`;
